### PR TITLE
[AIEX] AIECC-737 Combine more INSERT_SUBREG chains

### PR DESCRIPTION
**Pull request by gbossu**
Combine
```
    %1:ewl = COPY $wl2
    %2:vec256 = COPY $wh2
    %5:vec512 = REG_SEQUENCE %1, %subreg.sub_256_lo
    %6:vec512 = INSERT_SUBREG %5, %2, %subreg.sub_256_hi
```
into
```
    %1:ewl = COPY $wl2
    %2:vec256 = COPY $wh2
    %6:vec512 = REG_SEQUENCE %1, %subreg.sub_256_lo, %2, %subreg.sub_256_hi
```

### CODE REVIEW

File: llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp [hunk 1]
```diff
@@ -214,11 +214,23 @@ std::map<int, Register> collectSubregsChain(const MachineInstr &MI) {
   // E.g. %5:vec512 = INSERT_SUBREG %0, %1, %subreg.sub_256_hi
   //      %6:vec512 = INSERT_SUBREG %5, %2, %subreg.sub_256_lo
   std::function<void(const MachineInstr &)> Impl = [&](const MachineInstr &MI) {
+    if (MI.getOpcode() == TargetOpcode::REG_SEQUENCE) {
```
Comment (andcarminati) [inline llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp:217]: Hi, this implementation looks nice and clean, however this map (was here before) looks inverted. Every time we access it we reverse the results. Is it possible to use the same order as in REG_SEQUENCE? Is there any case were the same virtual subreg appears twice in a REQ_SEQUENCE / INSERT_SUBREG?

File: llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp [hunk 2]
```diff
@@ -214,11 +214,23 @@ std::map<int, Register> collectSubregsChain(const MachineInstr &MI) {
...
+
+    // Recursively follow INSERT_SUBREG chain
     assert(MI.getOpcode() == TargetOpcode::INSERT_SUBREG);
     Subregs.try_emplace(MI.getOperand(3).getImm(), MI.getOperand(2).getReg());
     MachineInstr &SrcMI = *MRI.getVRegDef(MI.getOperand(1).getReg());
     if (SrcMI.getParent() == MI.getParent() &&
-        SrcMI.getOpcode() == TargetOpcode::INSERT_SUBREG)
+        (SrcMI.getOpcode() == TargetOpcode::INSERT_SUBREG ||
+         SrcMI.getOpcode() == TargetOpcode::REG_SEQUENCE))
       Impl(SrcMI);
```
Comment (andcarminati) [inline llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp:234]: nit: CollectSubregsChainImpl

File: llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp (hunk 1 above)
Comment (gbossu) [reply llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp:217]: The reason it's ordered this way is exactly for the reason you mention: it's theoretically possible to have the same subreb multiple times in the sequence. In that case, I want to keep only the first instance I met when iterating bottom-up. I have a test for this particular case. I haven't seen it in real world examples, but you never know 🙈

File: llvm/test/CodeGen/AIE/aie2/GlobalISel/post-select-combine.mir [hunk 3]
```diff
@@ -158,3 +158,127 @@ body:             |
...
+---
+name:            combine_512_acc_with_reg_sequence
+alignment:       16
+legalized:       true
+regBankSelected: true
+selected:        true
+tracksRegLiveness: true
+body:             |
+  bb.0.entry:
+  liveins: $x0, $bml2, $bmh2
```
Comment (andcarminati) [inline llvm/test/CodeGen/AIE/aie2/GlobalISel/post-select-combine.mir]: nit: remove x0

File: llvm/test/CodeGen/AIE/aie2/GlobalISel/post-select-combine.mir [hunk 4]
```diff
@@ -158,3 +158,127 @@ body:             |
...
+---
+name:            combine_256_with_reg_sequence
+alignment:       16
+legalized:       true
+regBankSelected: true
+selected:        true
+tracksRegLiveness: true
+body:             |
+  bb.0.entry:
+  liveins: $x0, $wl2, $wh2
```
Comment (andcarminati) [inline llvm/test/CodeGen/AIE/aie2/GlobalISel/post-select-combine.mir]: nit: remove x0.

**Review by andcarminati — DISMISSED** [review]
LGTM. I left just 2 nits, but they are optional suggestions.

**Review by andcarminati — APPROVED** [review]

**Pull request merged**